static/js/jira-manual-app-creation.js

```diff
@@ -1,79 +1,148 @@
+	if (isUpdate) {
+		$(".jiraManualAppCreation__formNoFileUploaded").hide();
+		$(".jiraManualAppCreation__formFileUploaded").css('display', 'flex');
+		$("#privateKeyFile").attr("data-aui-validation-state", "valid");
+	}
+});
+
 AJS.$("#jiraManualAppCreation__form").on("aui-valid-submit", (event) => {
-  event.preventDefault();
-  const form = event.target;
-  const csrf = $("#_csrf").val();
-  const data = $(form).serializeObject();
-  const isUpdate = $('input[type=submit]').val() === "Update";
-
-  // Reading the content of the file
-  const file = $("#privateKeyFile")[0].files[0];
-  const reader = new FileReader();
-  reader.readAsDataURL(file);
-
-  reader.onload = () => {
-    data.privateKey = reader.result;
-
-    AP.context.getToken((token) => {
-      data.jwt = token;
-      data._csrf = csrf;
-      data.jiraHost = jiraHost;
-
-      if (isUpdate) {
-        // TODO: Do a put request to update the existing app
-      } else {
-         $.post("/jira/connect/enterprise/app", data, (response, _status, result) => {
-          if (result.status === 201) {
-            // TODO: This doesn't work, will be done in ARC-1565
-            const child = openChildWindow(`/session/github/${response.data.uuid}/configuration?ghRedirect=to`);
-            child.window.jiraHost = jiraHost;
-            child.window.jwt = token;
-          }
-        });
-      }
-    });
-  };
+	event.preventDefault();
+	const form = event.target;
+	const data = $(form).serializeObject();
+	const updateElement = document.getElementById("Update");
+	const isUpdate = updateELement && updateElement.innerText === "Update";
+	const uuid = $(event.target).data("app-uuid");
+	const appName = $(event.target).data("app-appname");
+	const existingPrivateKey = $(event.target).data("app-privatekey");
+
+	AP.context.getToken((token) => {
+		data.jwt = token;
+		data.jiraHost = jiraHost;
+
+		const renderedFilename = document.getElementById("jiraManualAppCreation__uploadedFile").innerText;
+		const isFileChanged = renderedFilename !== `${appName}.private-key.pem`;
+
+		if (isFileChanged || !isUpdate) {
+			// Reading the content of the file
+			const file = $("#privateKeyFile")[0].files[0];
+			const reader = new FileReader();
+			reader.readAsDataURL(file);
+
+			reader.onload = () => {
+				data.privateKey = reader.result;
+			};
+		} else {
+			data.privateKey = existingPrivateKey;
+		}
+
+		if (isUpdate) {
+			gitHubAppPutRequest(uuid, isUpdate, data);
+		} else {
+			$.post("/jira/connect/enterprise/app", data, (response, _status) => {
+				if (response.success) {
+					// TODO: This doesn't work, will be done in ARC-1565
+					const child = openChildWindow(`/session/github/${response.data.uuid}/configuration?ghRedirect=to`);
+					child.window.jiraHost = jiraHost;
+					child.window.jwt = token;
+				} else {
+					handleFormErrors(isUpdate);
+				}
+			});
+		}
+	});
+});
+
+$('#jiraManualAppCreation__uploadedFile').bind('DOMSubtreeModified', function () {
```

Comment:
Adding a flag when a file is uploaded. Because we aren't saving the name of the file, the best we can do to render a file name in editing mode is to use the format of <app-name>.private-key.pem. So if a user uploads a file, then goes back to edit the form, the 'incorrect' file naming will be confusing and may lead to customers thinking something has gone wrong. This notifies them of the file name formatting.

Reply:
/nit

I think I am not 100% following the code here. Too me seems like quite a lot of code going on for the `file name`. 

I wonder is that `file name` thing a UX requirement? Or is it just a technical issue (like the file input is mandatory)? 
It is just this `$( "input[type=text][name=gitHubAppName]" ).val().replace(/\s+/g, '-').toLowerCase()` quite brain energy consuming to me.

Reply:
What do you mean by 'file name thing'? I think you and @krazziekay had discussed rendering the file name in editing mode. As we don't save the file name of a stored private key, our only option is to use the format of <app-name>.private-key.pem. Jade has confirmed, from a UX standpoint, that is better that we have the file name rendered and don't force the user to have to reupload the private key any time they want to edit another app value.

This bind call detects a change in the element that handles displaying the file name. If a change occurs, we render a flag to inform the user that the private key file name will be stored in the format of <app-name>.private-key.pem. Otherwise, if a user saves their private key, which is in the format of  <app-name>.<created-at-date>.private-key.pem, and then returns to editing mode, the mismatched naming in editing mode will lead to confusion and may cause a customer to think the file didn't save correctly.  

What about `$( "input[type=text][name=gitHubAppName]" ).val().replace(/\s+/g, '-').toLowerCase()` is causing confusion? 
- `( "input[type=text][name=gitHubAppName]" ).val()` targets the input field with name `gitHubAppName` and retrieves the value e.g. My New GitHub App
- `.replace(/\s+/g, '-').toLowerCase()` replaces the spaces with hyphens and converts all casing to lower case e.g. my-new-github-app (as this is how the file name will be rendered in editing mode)

Reply:
Just my two cents, happy to have this thread close and move forward on this PR.

`from a UX standpoint, that is better that we have the file name rendered and` <-- what if, we don't display the `file name` at all, leave the file input box completely empty as new when editing existing app. User can choose to upload or leave it empty when click the save button. If user provides a new file, we override existing one in db. If leave it empty, we leave it as it is in db.

By doing above, do you think we can remove a whole bunch of code?

Reply:
Maybe but this was a UX request. I say we create a follow-up ticket and circle back on this one. I don't think it should be a blocker.

Reply:
Agree! closing this thread now.